src/instruction.rs

```diff
@@ -57,6 +57,90 @@ impl fmt::Display for ArithmeticOperator {
     }
 }
 
+#[derive(Clone, Debug, PartialEq)]
+pub enum BinaryOperand {
+    LiteralInteger(i64),
+    MemoryReference(MemoryReference),
+}
+
+impl fmt::Display for BinaryOperand {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match &self {
+            BinaryOperand::LiteralInteger(value) => write!(f, "{}", value),
+            BinaryOperand::MemoryReference(value) => write!(f, "{}", value),
+        }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum BinaryOperator {
+    And,
+    Ior,
+    Xor,
+}
+impl fmt::Display for BinaryOperator {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match &self {
+            BinaryOperator::And => write!(f, "AND"),
+            BinaryOperator::Ior => write!(f, "IOR"),
+            BinaryOperator::Xor => write!(f, "XOR"),
+        }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum UnaryOperator {
+    Neg,
+    Not,
+}
+
+impl fmt::Display for UnaryOperator {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match &self {
+            UnaryOperator::Neg => write!(f, "NEG"),
+            UnaryOperator::Not => write!(f, "NOT"),
+        }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum TernaryOperand {
+    LiteralInteger(i64),
+    LiteralReal(f64),
+    MemoryReference(MemoryReference),
+}
+
+impl fmt::Display for TernaryOperand {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match &self {
+            TernaryOperand::LiteralInteger(value) => write!(f, "{}", value),
+            TernaryOperand::LiteralReal(value) => write!(f, "{}", value),
+            TernaryOperand::MemoryReference(value) => write!(f, "{}", value),
+        }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum TernaryOperator {
```

Comment:
Not 100% on this - but like in your own comment, I wonder if this would be more intuitive as `ComparisonOperator`. The pattern of grouping multiple instructions by common function and syntax was meant to reduce duplicate code and match arms, but my goal was still to keep it intuitive to read in `Debug` format or in code. If I were looking to add an `LT` instruction to a program, I might not look for `TernaryInstruction`. This code is not just used to parse text, but also to build up programs at runtime.

Reply:
Yep, I think switching the name makes sense here. I will also do some thinking on addressing:

> That being said, I wonder if we'd want a suite of macros where the names are the Quil instruction names themselves, so there's a part of the code that maps 1<>1 with Quil, while not requiring all users of quil-rust to match against so many variants of Instruction (as they would have to if LT, GT, etc got their own variants)

Definitely could add something to help a user construct programs especially when using the grouped instructions I've included in this PR.